src/main/java/org/apache/mesos/scheduler/repair/RepairScheduler.java

```diff
@@ -0,0 +1,163 @@
+        newStopped = newStopped.stream().distinct().collect(Collectors.toList());
+        List<String> newFailed = terminatedTasks.stream()
+                .filter(it -> !failureMonitor.hasFailed(it))
+                .map(TaskInfo::getName)
+                .collect(Collectors.toList());
+
+        Optional<OfferRequirement> offerReq = Optional.empty();
+        boolean replacingStopped = false;
+
+        if (terminatedTasks.size() > 0) {
+            TaskInfo terminatedTask = terminatedTasks.get(new Random().nextInt(terminatedTasks.size()));
+            log.info("Found terminated task");
+            if (failureMonitor.hasFailed(terminatedTask)) {
+                log.info("Purging terminated task");
+                failureListener.taskFailed(terminatedTask.getTaskId());
+            } else {
+                log.info("Relaunching terminated task in place");
+                offerReq = Optional.of(offerReqProvider.getReplacementOfferRequirement(terminatedTask));
+            }
+        } else {
+            log.info("Launching a stopped task");
+            replacingStopped = true;
+            offerReq = offerReqProvider.maybeGetNewOfferRequirement(targetConfigName, block);
+        }
+
+        if (offerReq.isPresent() && launchConstrainer.canLaunch(offerReq.get())) {
+            log.info("Preparing to launch task");
+            OfferEvaluator offerEvaluator = new OfferEvaluator();
+            List<OfferRecommendation> recommendations = offerEvaluator.evaluate(offerReq.get(), offers);
+            if (!recommendations.isEmpty()) {
+                // we're replacing some task(s)
+                Set<String> relaunchedTasks = recommendations.stream()
+                        .filter(it -> it.getOperation().hasLaunch())
+                        .flatMap(it -> it.getOperation().getLaunch().getTaskInfosList().stream())
+                        .map(TaskInfo::getName)
+                        .collect(Collectors.toSet());
+                if (replacingStopped) {
+                    // we'll filter their names out of the stopped pool
+                    newStopped = newStopped.stream()
+                            .filter(it -> !relaunchedTasks.contains(it))
+                            .collect(Collectors.toList());
+                } else {
+                    // we'll filter their names out of the failed pool
+                    newFailed = newFailed.stream()
+                            .filter(it -> !relaunchedTasks.contains(it))
+                            .collect(Collectors.toList());
+                }
+            }
+            List<Operation> launchOperations = recommendations.stream()
+                    .map(OfferRecommendation::getOperation)
+                    .filter(Operation::hasLaunch)
+                    .collect(Collectors.toList());
+            if (launchOperations.size() > 1) {
+                throw new RuntimeException("Repairs should only launch one task at a time, found "
```

Comment:
I can switch to checked exceptions, but they haven't been popular for usability reasons for many years. I typically throw RuntimeExceptions when I would like to use an `assert`, but of course `assert`s can be disabled, and this is a critical check.

The point here is to help developers catch incorrect implementations of the provider at dev time. I don't think that requiring try/catch around callsites of this method will help, since this is meant to detect bad implementations early on in the development process.